File: solutions/python/05/lis_fast/main.py
import bisect


def lis_length(a: list) -> int:
    tails = []
    for x in a:
        i = bisect.bisect_left(tails, x)
        if i == len(tails):
            tails.append(x)
        else:
            tails[i] = x
    return len(tails)


def main() -> None:
    print(lis_length([3, 1, 4, 2, 5]))


if __name__ == "__main__":
    main()

File: solutions/go/05/lis_fast/main.go
package main

import "fmt"

func lowerBound(a []int, x int) int {
	lo, hi := 0, len(a)
	for lo < hi {
		mid := (lo + hi) / 2
		if a[mid] < x {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	return lo
}

func lisLength(a []int) int {
	tails := []int{}
	for _, x := range a {
		i := lowerBound(tails, x)
		if i == len(tails) {
			tails = append(tails, x)
		} else {
			tails[i] = x
		}
	}
	return len(tails)
}

func main() {
	fmt.Println(lisLength([]int{3, 1, 4, 2, 5}))
}
